ext/drmgr/drmgr.c

```diff
@@ -3313,6 +3319,45 @@ drmgr_in_emulation_region(void *drcontext, OUT const emulated_instr_t **emulatio
     return true;
 }
 
+DR_EXPORT
+instr_t *
+drmgr_orig_app_instr_for_fetch(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, our_tls_idx);
+    if (drmgr_current_bb_phase(drcontext) != DRMGR_PHASE_INSERTION)
+        return NULL;
+    const emulated_instr_t *emulation;
+    if (drmgr_in_emulation_region(drcontext, &emulation)) {
+        if (TEST(DR_EMULATE_FIRST_INSTR, emulation->flags)) {
+            return emulation->instr;
+        } // Else skip further instr fetches until outside emulation region.
+    } else if (instr_is_app(pt->insertion_instr)) {
```

Comment:
About skipping meta instrs in insertion phase: For a client that changes app behaviour (not expansion related) in app2app phase, I guess it can have some drreg use in app2app that adds meta instrs right? In this case, it may want to look at these meta instrs in the insertion phase. Or is such a client expected to set translation for these instrs then?

Reply:
An observational client should be able to completely ignore meta instructions.  Meta instructions are not supposed to affect application state.

Reply:
Okay. So the app2app phase is expected to set translation for app-state-affecting instrs then, so that they are not considered meta.

Reply:
Not just translation: it should not call instr_set_meta (or indirectly via instrlist_meta_preinsert) on instrs that are supposed to be viewed as the app instruction stream by clients.